Add authentication

toban opened the pull request.

lucaswerkmeister commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -34,16 +37,21 @@ class EditEndpoint extends SimpleHandler {
 	private $simplePutStrategy;
 
+	/** @var User */
+	private $user;
```
Is this supposed to be a `User` or a `UserIdentity`? It’s `UserIdentity` in the constructor but `User` in the phpdoc comment.

lucaswerkmeister commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -141,17 +157,15 @@ public function run() {
 				EDIT_NEW,
-				// TODO actually do a token check?
-				false
+				$this->user->getEditToken()
```
I don’t think this is the right way to do the token check – we’re getting the user’s token and then asking `EditEntity` to compare it to the user’s token, which should always succeed. The token should come from the request parameters, if I understand correctly.

lucaswerkmeister replied:
Hm, in master of MediaWiki core, `UpdateHandler` and `EditHandler::getActionModuleToken()` have some interesting related code – apparently session providers can declare whether they’re potentially vulnerable to CSRF or not, and depending on that the `token` parameter is not required. (In fact, if the session declares itself safe but the request still includes a token, an error is raised, which strikes me as overly hostile to API consumers – but whatever.)

But that code is currently in `EditHandler`, a subclass of `ActionModuleBasedHandler`, so we probably couldn’t directly use it even if we were targeting MediaWiki master (rather than 1.35).

toban replied:
Ok, have a look at https://github.com/wmde/WikibaseReconcileEdit/pull/35/commits/2f1132640bf84050be6050160d0143894462899b

It switches these parameters from being passed in as post parameters but rather all together in the body. Made the tests and the code to look a bit more similar to what is being done in UpdateHandler.

lucaswerkmeister commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -197,19 +199,31 @@ public function needsWriteAccess() {
+		return new JsonBodyValidator( [
 			'entity' => [
-				self::PARAM_SOURCE => 'post',
+				self::PARAM_SOURCE => 'body',
```
This isn’t actually used by `JsonBodyValidator` as far as I can tell, and `'body'` is not one of the allowed values according to the constant documentation, so I think we should just remove these lines.

toban replied:
done, added a test this validation too.

lucaswerkmeister commented on includes/MediaWiki/Request/EditRequest.php:
```diff
@@ -17,23 +17,29 @@ class EditRequest {
 		$this->reconcilePropertyId = $reconcilePropertyId;
 		$this->entity = $entity;
 		$this->otherItems = $otherItems;
+		$this->token = $token;
```
(I’d prefer to move this up one line, so that the field, constructor params, assignments and getters are all in the same order.)

toban replied:
done

lucaswerkmeister commented on includes/MediaWiki/Request/EditRequestParser.php:
```diff
@@ -39,11 +38,9 @@ public function __construct(
+
 		if ( !is_array( $reconcile ) ) {
 			throw new LocalizedHttpException(
 				MessageValue::new( 'wikibasereconcileedit-editendpoint-invalid-reconcile-json' ),
```
This should probably be a different error message, since the JSON syntax is now checked by `JsonBodyValidator`.

toban replied:
done

lucaswerkmeister commented on includes/MediaWiki/Request/EditRequestParser.php:
```diff
@@ -84,10 +81,7 @@ public function parseRequestInterface( RequestInterface $request ): EditRequest
+		$entity = $requestBody['entity'] ?: null;
 
 		if (
 			// TODO separate error message for invalid entity JSON?
```
This TODO comment is also obsolete in that form, I think.

lucaswerkmeister commented on includes/MediaWiki/Request/EditRequestParser.php:
```diff
@@ -118,7 +112,8 @@ public function parseRequestInterface( RequestInterface $request ): EditRequest
 			$reconcilePropertyId,
 			$inputEntity,
-			$otherItems
+			array_key_exists( 'token', $requestBody ) ? $requestBody['token'] : '',
```
This shouldn’t be necessary, I think, since we already specify a default value for the `JsonBodyValidator`.

lucaswerkmeister commented on tests/phpunit/unit/MediaWiki/Request/EditRequestParserTest.php:
```diff
@@ -164,14 +164,14 @@ public function provideUnsupportedEntityVersion(): iterable {
 
 	/** @dataProvider provideInvalidReconcileJson */
 	public function testParseRequestInterface_invalidReconcileJson( string $reconcile ): void {
```
This test doesn’t make sense anymore – the request parser no longer decodes JSON.

toban replied:
done

lucaswerkmeister commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -73,16 +72,25 @@ public static function factory(
 
 	public function run() {
-		if ( !$this->user->isRegistered() ) {
+		// Rest Validator returns null if submitted as form
```
Hm, are you sure? From looking at the code, it seems like `run()` should never even run if validation fails, at least when using MediaWiki’s `Router`.

toban replied:
It exists before the body validator is run here https://github.com/wikimedia/mediawiki/blob/REL1_35/includes/Rest/Validator/Validator.php#L163

lucaswerkmeister replied:
:roll_eyes: I see, thanks.

lucaswerkmeister commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -73,16 +72,25 @@ public static function factory(
+		$requestBody = $this->getValidatedBody();
+		if ( $requestBody === null ) {
+			throw new LocalizedHttpException(
+				MessageValue::new( 'wikibasereconcileedit-editendpoint-invalid-reconcile-json' ),
```
I don’t think this is the right message, since it’s not really the reconcile JSON, specifically, that’s invalid.

toban replied:
done

lucaswerkmeister commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -203,9 +211,9 @@ public function needsWriteAccess() {
-				415,
-				[ 'content_type' => $contentType ]
+			throw new LocalizedHttpException(
+				MessageValue::new( 'wikibasereconcileedit-editendpoint-invalid-reconcile-json' ),
```
This also doesn’t seem like the right use case for that message?

toban replied:
done

lucaswerkmeister commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -57,14 +65,32 @@ public static function factory(
+			);
+		}
+
+		if ( !$this->user->isRegistered() || !$this->user->matchEditToken( $requestBody['token'] ) ) {
```
If we’re checking the edit token here anyways (which I think is a good idea, because it happens before we start handing out new item IDs), then we could probably continue to pass `false` into `attemptSave()`? Or is it useful to do the token check twice?

toban replied:
Feels a bit backwards not using the token when saving if it was included for this purpose?

lucaswerkmeister replied:
Fair enough ^^

toban committed:
Add Authentication

- adds required token parameter
- removes some seemingly uncessary json encoding
- EditRequestParser no longer parses a requestInterface but the
validated body from the handler.
- passes the token and other parameters in the request body like other
MW simpleHandlers do.

toban committed:
Add example python scripts and some documentation

toban committed:
Update test system OAuth to be managed by sysop

lucaswerkmeister commented on i18n/en.json:
```diff
@@ -15,7 +15,10 @@
 	"wikibasereconcileedit-editendpoint-reconciliation-property-missing-in-statements": "Entity must have at least one statement for the reconciliation Property",
 	"wikibasereconcileedit-editendpoint-invalid-reconciliation-statement-type": "Reconciliation statement must contain a value",
+	"wikibasereconcileedit-editendpoint-invalid-request-body": "Invalid or empty JSON request body",
+	"wikibasereconcileedit-editendpoint-invalid-content-type": "Content-type should be application/json",
```
Hm, maybe the expected content type should be a message parameter?

toban replied:
sure!

lucaswerkmeister commented on i18n/en.json:
```diff
@@ -15,7 +15,10 @@
 	"wikibasereconcileedit-minimaliteminput-required-keys": "Statements missing '$1' key",
-	"wikibasereconcileedit-fullwikibaseiteminput-unsupported-type": "The only supported entity {{PLURAL:$2|type is|types are}} $1."
+	"wikibasereconcileedit-fullwikibaseiteminput-unsupported-type": "The only supported entity {{PLURAL:$2|type is|types are}} $1.",
+	"wikibasereconcileedit-unauthorized-access": "Not logged in."
```
Is it intentional that this doesn’t mention the possibility of an invalid CSRF token?

toban replied:
I updated to be a bit more specific what could be wrong.

lucaswerkmeister commented on README.md:
```diff
@@ -104,6 +104,19 @@ Edit payload `0.0.1/full` should look like this:
+
+OAuth requires the [OAuth Extension](https://www.mediawiki.org/wiki/Extension:OAuth) to be installed and a OAuth consumer created via the `wiki/Special:OAuthConsumerRegistration/propose` page.
+
+Follow the instructions [here]((https://www.mediawiki.org/wiki/OAuth/Owner-only_consumers)) on how to create a OAuth owner-only consumer.
```
There’s an extra set of `()` here, which apparently causes GitHub to not render this as a link at all?

toban replied:
done

lucaswerkmeister commented on README.md:
```diff
@@ -104,6 +104,19 @@ Edit payload `0.0.1/full` should look like this:
+
+For an example on how to request a edit token and make an authenticated request using OAuth see the [example python script](example/oauth/main.py).
+
+**Note: OAuth owner-only consumers has not been confirmed to work without https**
```
Is this still the case?

toban replied:
I wasn't able to get it working :|

lucaswerkmeister replied:
I tested it locally and it worked fine for me.

toban committed:
CSRF token should not be mandatory

- OAuth sessionprovider will be considered safe against csrf
- Inject sessionprovider into EditEndpoint

lucaswerkmeister commented on tests/phpunit/unit/MediaWiki/Request/EditRequestParserTest.php:
```diff
@@ -122,14 +123,14 @@ public function testParseRequestInterface_good(): void {
 	}
```
There should probably be an assertion here about the `$editRequest->token()`? (Unless you want to squash-merge the PR, in which case it’s moot because the `token()` goes away again anyways ^^)

lucaswerkmeister replied:
Done.

lucaswerkmeister commented on i18n/qqq.json:
```diff
@@ -15,7 +15,11 @@
 	"wikibasereconcileedit-editendpoint-reconciliation-property-missing-in-statements": "Error returned by the API if the “entity” request parameter doesn't contain a statement that uses the reconciliation property",
 	"wikibasereconcileedit-editendpoint-invalid-reconciliation-statement-type": "Error returned by the API if the “entity” request parameter if the reconciliation statement's mainsnak is not of type value",
+	"wikibasereconcileedit-editendpoint-invalid-request-body": "Error returned by the API if the JSON request body is invalid or empty.",
+	"wikibasereconcileedit-editendpoint-invalid-content-type": "Error returned by the API if the request content type is not application/json",
```
This should probably explain what `$1` is and not mention `application/json`.

lucaswerkmeister replied:
Done.

lucaswerkmeister commented on tests/phpunit/integration/MediaWiki/Api/EditEndpointTest.php:
```diff
@@ -302,4 +304,10 @@ private function countItemsInDatabase(): int {
 		);
 	}
 
+	private function getMockSessionProvider( bool $safeAgainstCSRF ): SessionProviderInterface {
```
Hm, there’s no test that passes `false` here?

lucaswerkmeister replied:
Added a test, I hope it looks okay.

toban replied:
nice :+1:

toban commented:
LGTM!
